metrics-core/src/lib.rs

```diff
@@ -32,24 +32,207 @@
 //!
 //! Histograms are a convenient way to measure behavior not only at the median, but at the edges of
 //! normal operating behavior.
+#![deny(missing_docs)]
 use futures::future::Future;
 use std::borrow::Cow;
+use std::fmt;
+use std::slice::Iter;
 use std::time::Duration;
 
-/// An optimized metric key.
+/// An allocation-optimized string.
 ///
-/// As some metrics might be sent at high frequency, it makes no sense to constantly allocate and
-/// reallocate owned [`String`]s when a static [`str`] would suffice.  As we don't want to limit
-/// callers, though, we opt to use a copy-on-write pointer -- [`Cow`] -- to allow callers
-/// flexiblity in how and what they pass.
-pub type Key = Cow<'static, str>;
+/// We specify `ScopedString` to attempt to get the best of both worlds: flexibility to provide a
+/// static or dynamic (owned) string, while retaining the performance benefits of being able to
+/// take ownership of owned strings and borrows of completely static strings.
+pub type ScopedString = Cow<'static, str>;
 
-/// A value which can be converted into a nanosecond representation.
+/// A key/value pair used to further describe a metric.
+#[derive(PartialEq, Eq, Hash, Clone, Debug)]
+pub struct Label(ScopedString, ScopedString);
+
+impl Label {
+    /// Creates a `Label` from a key and value.
+    pub fn from_parts<K, V>(key: K, value: V) -> Self
+    where
+        K: Into<ScopedString>,
+        V: Into<ScopedString>,
+    {
+        Label(key.into(), value.into())
+    }
+
+    /// The key of this label.
+    pub fn key(&self) -> &str {
+        self.0.as_ref()
+    }
+
+    /// The value of this label.
+    pub fn value(&self) -> &str {
+        self.1.as_ref()
+    }
+
+    /// Consumes this `Label`, returning the key and value.
+    pub fn into_parts(self) -> (ScopedString, ScopedString) {
+        (self.0, self.1)
+    }
+}
+
+/// A metric key.
+///
+/// A key always includes a name, but can optional include multiple labels used to further describe
+/// the metric.
+#[derive(PartialEq, Eq, Hash, Clone, Debug)]
+pub struct Key {
+    name: ScopedString,
+    labels: Vec<Label>,
+}
+
+impl Key {
+    /// Creates a `Key` from a name.
+    pub fn from_name<N>(name: N) -> Self
+    where
+        N: Into<ScopedString>,
+    {
+        Key {
+            name: name.into(),
+            labels: Vec::new(),
+        }
+    }
+
+    /// Creates a `Key` from a name and vector of `Label`s.
+    pub fn from_name_and_labels<N, L>(name: N, labels: L) -> Self
+    where
+        N: Into<ScopedString>,
+        L: Into<Vec<Label>>,
```

Comment:
Seems unlikely that it would be usefull? Maybe `IntoIterator<Item = Label>`, followed by a collect would be more usefull here. Can't think of anything that implements `Into<Vec<T>>` from the top of my head.

Reply:
Yeah, this is going to change once I redo the impls for getting the vector of labels.